securedrop/source_app/decorators.py

```diff
@@ -1,19 +1,28 @@
 from typing import Any
 
-from flask import redirect, url_for, request
+from flask import redirect, url_for, request, session
 from functools import wraps
 
 from typing import Callable
 
-from source_app.utils import logged_in
+from source_app.utils import clear_session_and_redirect_to_logged_out_page
+from source_app.session_manager import SessionManager, UserNotLoggedIn, \
+    UserSessionExpired, UserHasBeenDeleted
 
 
 def login_required(f: Callable) -> Callable:
     @wraps(f)
     def decorated_function(*args: Any, **kwargs: Any) -> Any:
-        if not logged_in():
-            return redirect(url_for('main.login'))
-        return f(*args, **kwargs)
+        try:
+            logged_in_source = SessionManager.get_logged_in_user()
+
+        except (UserSessionExpired, UserHasBeenDeleted):
+            return clear_session_and_redirect_to_logged_out_page(session)
+
+        except UserNotLoggedIn:
+            return redirect(url_for("main.login"))
+
+        return f(*args, **kwargs, logged_in_source=logged_in_source)
```

Comment:
The decorator directly passes the source user to the Flask route/code, so using flask.g or session (and checking that they are in the right "state") is no longer needed.